javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -0,0 +1,164 @@
+
+    private static final BitMappedTrie<?> EMPTY = new BitMappedTrie<>(emptyArray(), 0, 0, 0);
+
+    @SuppressWarnings("unchecked")
+    static <T> BitMappedTrie<T> emptyTree() { return (BitMappedTrie<T>) EMPTY; }
+
+    private final Object[] array;
+    private final int offset, length;
+    private final int depthShift;
+
+    BitMappedTrie(Object[] array, int offset, int length, int depthShift) {
+        this.array = array;
+        this.offset = offset;
+        this.length = length;
+        this.depthShift = depthShift;
+
+        assert length() <= treeSize(branchingFactor(), depthShift);
+    }
+
+    private static int treeSize(int branchCount, int depthShift) {
+        final int fullBranchSize = 1 << depthShift;
+        return branchCount * fullBranchSize;
+    }
+
+    BitMappedTrie<T> prependLeaf(T leading) {
+        final int newSize = length() + 1;
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(leading), offset, newSize, depthShift);
+        } else {
+            Object[] array = this.array;
+            int depthShift = this.depthShift, offset = this.offset;
+            if (isFullLeft()) {
+                array = copyUpdate(emptyArray(), branchingFactor() - 1, array);
+                depthShift += BRANCHING_BASE;
+                offset = treeSize(branchingFactor() - 1, depthShift);
+            }
+
+            offset -= 1;
+            array = modifyLeaf(array, depthShift, offset, NODE_MODIFIER, LEAF_MODIFIER.apply(leading));
+            return new BitMappedTrie<>(array, offset, newSize, depthShift);
+        }
+    }
+    private boolean isFullLeft() { return offset == 0; }
+
+    BitMappedTrie<T> appendLeaf(T trailing) {
+        final int newSize = length() + 1;
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(trailing), offset, newSize, depthShift);
+        } else {
+            Object[] array = this.array;
+            int depthShift = this.depthShift;
+            if (isFullRight(newSize)) {
+                array = asArray(array);
+                depthShift += BRANCHING_BASE;
+            }
+
+            array = modifyLeaf(array, depthShift, offset + length(), NODE_MODIFIER, LEAF_MODIFIER.apply(trailing));
+            return new BitMappedTrie<>(array, offset, newSize, depthShift);
+        }
+    }
+    private boolean isFullRight(int newSize) { return (offset + newSize) > treeSize(branchingFactor(), depthShift); }
+
+    private static final NodeModifier NODE_MODIFIER = (o, i) -> copy(o, i + 1);
+    private static final Function<Object, NodeModifier> LEAF_MODIFIER = element -> ((o, i) -> copyUpdate(o, i, element));
+
+    BitMappedTrie<T> update(int index, T element) {
+        final Object[] root = modifyLeaf(array, depthShift, offset + index, NODE_MODIFIER, LEAF_MODIFIER.apply(element));
+        return new BitMappedTrie<>(root, offset, length(), depthShift);
+    }
+
+    BitMappedTrie<T> drop(int n) {
```

Comment:
`drop` and `tail` could be optimized to only modifying the tree, if anything other than the leaf should be changed (i.e. something like `if (n - offset > branchingFactor())`)